File: backend/helper_files/CommentsHelper.js
import sqlite3 from 'sqlite3';
const dbFile = 'instagram_data.sqlite';
import dotenv from 'dotenv';
import { pipeline } from "@xenova/transformers";

dotenv.config({ path: './.env' });
const instaUserId = process.env.INSTA_USER_ID; 

/**
 * Retrieves all comments from the SQLite database for a given Instagram user.
 * @returns {Promise<string[]>} A promise that resolves to an array of comment texts.
 */
async function getAllComments() {
  const db = new sqlite3.Database(dbFile);
  const query = `SELECT comments FROM user${instaUserId} WHERE comments IS NOT NULL;`;
  
  return new Promise((resolve, reject) => {
    db.all(query, [], (err, rows) => {
      if (err) {
        console.error("Database error:", err);
        reject(err);
        return;
      }
      
      var comments = [];
      for (const row of rows) {
        var commentsSection = JSON.parse(row.comments);
        for (const post of commentsSection) {
          var comment = post.text;
          comments.push(comment);
        }
      }
      
      db.close();
      resolve(comments);
    });
  });
}

/**
 * Classifies comments using a pre-trained transformer model for sentiment analysis.
 * @returns {Promise<Object[]>} A promise that resolves to an array of sentiment statistics.
 */
async function classifyComments() {
  const classifier = await pipeline("sentiment-analysis", "Xenova/distilbert-base-uncased-finetuned-sst-2-english");
  var comments = await getAllComments();
  const sentimentStats = {
      POSITIVE: { label: "POSITIVE", count: 0, totalScore: 0 },
      NEGATIVE: { label: "NEGATIVE", count: 0, totalScore: 0 },
      NEUTRAL: { label: "NEUTRAL", count: 0, totalScore: 0 }
  };

  for (const comment of comments) {
      const [result] = await classifier(comment);
      const { label, score } = result;

      if (label in sentimentStats) {
          sentimentStats[label].count += 1;
          sentimentStats[label].totalScore += score;
      } else {
          sentimentStats.NEUTRAL.count += 1;
          sentimentStats.NEUTRAL.totalScore += score;
      }
  }

  return Object.values(sentimentStats).map(({ label, count, totalScore }) => ({
      label,
      count,
      avgScore: count > 0 ? (totalScore / count).toFixed(4) : 0
  }));
}

export { classifyComments };

File: backend/helper_files/CommentsHelper.py
from transformers import pipeline
import sqlite3
import json
import os
from dotenv import load_dotenv

load_dotenv()
INSTA_USER_ID = os.getenv('INSTA_USER_ID')

# Initialize the emotion detection model using a pre-trained DistilRoBERTa-based classifier
emotion_model = pipeline("text-classification", model="j-hartmann/emotion-english-distilroberta-base")

def get_all_comments(db_file):
    """
    Retrieves all comments from the SQLite database for a given Instagram user.

    Parameters:
        db_file (str): The path to the SQLite database file.
        insta_user_id (str): The Instagram user ID whose comments should be retrieved.

    Returns:
        list: A list of comment texts.

    Process:
    1. Establish a connection to the SQLite database.
    2. Execute an SQL query to fetch non-null comments from the user's table.
    3. Parse the JSON-formatted comments stored in the database.
    4. Extract and return individual comment texts.
    5. Handle database errors gracefully and ensure the connection is closed.
    """
    conn = sqlite3.connect(db_file)
    cursor = conn.cursor()    
    query = f"SELECT comments FROM user{INSTA_USER_ID} WHERE comments IS NOT NULL;"
    
    try:
        cursor.execute(query)
        rows = cursor.fetchall()
        comments = []
        for row in rows:
            comments_section = json.loads(row[0])  
            for post in comments_section:
                comment = post['text']
                comments.append(comment)
        
        return comments
    except sqlite3.Error as err:
        print("Database error:", err)
        return []
    finally:
        conn.close()

def detect_emotions_for_comments_from_db():
    """
    Detects emotions for all comments retrieved from the database.

    Returns:
        list: A list of detected emotion labels for each comment.

    Process:
    1. Calls `get_all_comments()` to fetch comments from the database.
    2. Uses the pre-trained DistilRoBERTa emotion detection model to classify each comment.
    3. Extracts the predicted emotion label for each comment and returns them as a list.
    4. If no comments are found, returns an empty list.
    """
    comments = get_all_comments('instagram_data.sqlite')
    if not comments:
        return []
    
    emotions = [emotion_model(comment)[0]['label'] for comment in comments]
    
    return emotions
